src/OpenSearch/TransportFactory.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace OpenSearch;
+
+use Http\Discovery\Psr17FactoryDiscovery;
+use Http\Discovery\Psr18ClientDiscovery;
+use OpenSearch\Serializers\SerializerInterface;
+use OpenSearch\Serializers\SmartSerializer;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface as PsrRequestFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+
+/**
+ * Creates a PSR transport falling back to a discovery mechanism if properties are not specified.
+ */
+class TransportFactory
+{
+    protected ?PsrRequestFactoryInterface $psrRequestFactory = null;
+
+    protected ?StreamFactoryInterface $streamFactory = null;
+
+    protected ?UriFactoryInterface $uriFactory = null;
+
+    protected ?SerializerInterface $serializer = null;
+
+    protected ?RequestFactoryInterface $requestFactory = null;
+
+    protected ?ClientInterface $httpClient = null;
```

Comment:
I would suggest making those properties private, to simplify maintenance. Implementing a BC layer for protected properties is almost impossible in PHP (until we can target PHP 8.4+ which gives us property hooks for that).
As all properties have public setters and protected getters, this would not hurt extensibility for child classes.